### Imports

In [1]:
import pandas as pd
import numpy as np


In [2]:
retail_sales_data=pd.read_csv('retail_store_sales.csv')

## Exploration

#### Columns from the retail sales data

In [3]:
column_names = retail_sales_data.columns
print(column_names)

Index(['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit',
       'Quantity', 'Total Spent', 'Payment Method', 'Location',
       'Transaction Date', 'Discount Applied'],
      dtype='object')


In [4]:
retail_sales_data.sample(n=20)

Transaction ID Customer ID                            Category  \
1338     TXN_1966346     CUST_23                          Patisserie   
9434     TXN_4949721     CUST_08  Computers and electric accessories   
6755     TXN_5819222     CUST_14                       Milk Products   
10337    TXN_9659189     CUST_13  Computers and electric accessories   
3260     TXN_9430988     CUST_13                       Milk Products   
1970     TXN_4621950     CUST_04  Computers and electric accessories   
6397     TXN_8050790     CUST_05  Computers and electric accessories   
1185     TXN_6781800     CUST_24                           Furniture   
2723     TXN_1153995     CUST_03       Electric household essentials   
2336     TXN_4123109     CUST_24                           Beverages   
2693     TXN_4240172     CUST_19                            Butchers   
535      TXN_2263002     CUST_09                          Patisserie   
3389     TXN_8737583     CUST_01                       Milk Products   
6081     TXN_5606325     CUST_23                           Beverages   
1422     TXN_8071580     CUST_20                       Milk Products   
4078     TXN_7017414     CUST_20                           Beverages   
2324     TXN_5679757     CUST_19                            Butchers   
3948     TXN_3962622     CUST_20                       Milk Products   
9067     TXN_1175373     CUST_13                       Milk Products   
71       TXN_7363165     CUST_25                           Beverages   

               Item  Price Per Unit  Quantity  Total Spent  Payment Method  \
1338    Item_11_PAT            20.0      10.0        200.0            Cash   
9434            NaN            23.0       NaN          NaN     Credit Card   
6755   Item_20_MILK            33.5       1.0         33.5     Credit Card   
10337           NaN            33.5       NaN          NaN     Credit Card   
3260            NaN            36.5       NaN          NaN            Cash   
1970    Item_12_CEA            21.5       8.0        172.0  Digital Wallet   
6397    Item_11_CEA            20.0       6.0        120.0            Cash   
1185    Item_19_FUR            32.0       8.0        256.0            Cash   
2723            NaN            23.0       NaN          NaN     Credit Card   
2336            NaN             NaN       2.0         19.0  Digital Wallet   
2693    Item_20_BUT            33.5       4.0        134.0     Credit Card   
535     Item_16_PAT            27.5       9.0        247.5  Digital Wallet   
3389   Item_13_MILK            23.0       7.0        161.0     Credit Card   
6081     Item_8_BEV            15.5       1.0         15.5  Digital Wallet   
1422   Item_22_MILK            36.5       1.0         36.5  Digital Wallet   
4078    Item_21_BEV            35.0       3.0        105.0            Cash   
2324    Item_22_BUT            36.5       8.0        292.0  Digital Wallet   
3948   Item_11_MILK            20.0       5.0        100.0     Credit Card   
9067   Item_24_MILK            39.5       3.0        118.5     Credit Card   
71       Item_4_BEV             9.5       7.0         66.5  Digital Wallet   

       Location Transaction Date Discount Applied  
1338   In-store       2022-11-29              NaN  
9434   In-store       2022-04-28             True  
6755   In-store       2022-06-10             True  
10337    Online       2024-06-20              NaN  
3260   In-store       2024-05-31             True  
1970     Online       2025-01-08            False  
6397   In-store       2024-06-05              NaN  
1185     Online       2024-09-26             True  
2723   In-store       2022-02-12            False  
2336   In-store       2025-01-12             True  
2693     Online       2022-07-26             True  
535      Online       2022-06-10            False  
3389   In-store       2023-02-21              NaN  
6081     Online       2024-08-09              NaN  
1422   In-store       2022-09-03            False  
4078   In-store       

In [5]:
print(retail_sales_data.describe(include='all'))

       Transaction ID Customer ID                       Category        Item  \
count           12575       12575                          12575       11362   
unique          12575          25                              8         200   
top       TXN_2407494     CUST_05  Electric household essentials  Item_2_BEV   
freq                1         544                           1591         126   
mean              NaN         NaN                            NaN         NaN   
std               NaN         NaN                            NaN         NaN   
min               NaN         NaN                            NaN         NaN   
25%               NaN         NaN                            NaN         NaN   
50%               NaN         NaN                            NaN         NaN   
75%               NaN         NaN                            NaN         NaN   
max               NaN         NaN                            NaN         NaN   

        Price Per Unit      Quantity   

In [6]:
retail_sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [7]:
retail_sales_data.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

#### what I have understood
- Categories - 8
- Items - 200/8 = 25, 
- which means there are 25 items under each category(8), which makes 200 unique items.
Price Per unit - 

#### Here are the categories 

In [8]:
retail_sales_data['Category'].unique()

array(['Patisserie', 'Milk Products', 'Butchers', 'Beverages', 'Food',
       'Furniture', 'Electric household essentials',
       'Computers and electric accessories'], dtype=object)

In [9]:
items_list = retail_sales_data['Item'].unique()

In [10]:
len(items_list) 

201

In [11]:
price_per_unit_list = (retail_sales_data['Price Per Unit'].unique())
print(price_per_unit_list)
print(len(price_per_unit_list))

[18.5 29.  21.5 27.5 12.5  nan  5.  33.5 36.5  8.   6.5 39.5 24.5 23.
 35.  14.   9.5 41.  20.  38.  15.5 11.  32.  26.  30.5 17. ]
26


In [12]:
item_price_summary = retail_sales_data.groupby('Item')['Price Per Unit'].unique().reset_index()

item_price_summary['Price Per Unit'] = item_price_summary['Price Per Unit'].str[0]

print(item_price_summary)

             Item  Price Per Unit
0     Item_10_BEV            18.5
1     Item_10_BUT            18.5
2     Item_10_CEA            18.5
3     Item_10_EHE            18.5
4    Item_10_FOOD            18.5
..            ...             ...
195    Item_9_EHE            17.0
196   Item_9_FOOD            17.0
197    Item_9_FUR            17.0
198   Item_9_MILK            17.0
199    Item_9_PAT            17.0

[200 rows x 2 columns]


In [13]:
unique_price_with_items = item_price_summary.groupby('Price Per Unit')

In [ ]:
for price, group in unique_price_with_items:
    print(group.sample(n=1))


           Item  Price Per Unit
84  Item_1_FOOD             5.0
           Item  Price Per Unit
143  Item_2_PAT             6.5
            Item  Price Per Unit
150  Item_3_MILK             8.0
           Item  Price Per Unit
159  Item_4_PAT             9.5
           Item  Price Per Unit
163  Item_5_EHE            11.0
           Item  Price Per Unit
168  Item_6_BEV            12.5
           Item  Price Per Unit
177  Item_7_BUT            14.0
           Item  Price Per Unit
186  Item_8_CEA            15.5
           Item  Price Per Unit
195  Item_9_EHE            17.0
          Item  Price Per Unit
2  Item_10_CEA            18.5
            Item  Price Per Unit
12  Item_11_FOOD            20.0
           Item  Price Per Unit
19  Item_12_EHE            21.5
            Item  Price Per Unit
28  Item_13_FOOD            23.0
           Item  Price Per Unit
35  Item_14_EHE            24.5
           Item  Price Per Unit
42  Item_15_CEA            26.0
           Item  Price Per Unit
55  

As you can see, Inspite of the category each item have unique price, from 5 to 41

- item 1 = 5.0
- item 2 = 6.5
- item 3 = 8.0
- item 4 = 9.5
- item 5 = 11.0
- item 6 = 12.5
- item 7 = 14.0
- item 8 = 15.5
- item 9 = 17.0
- item 10 = 18.5
- item 11 = 20.0
- item 12 = 21.5
- item 13 = 23.0
- item 14 = 24.5
- item 15 = 26.0 
- item 16 = 27.5
- item 17 = 29.0
- item 18 = 30.5
- item 19 = 32.0
- item 20 = 33.5
- item 21 = 35.0
- item 22 = 36.5
- item 23 = 38.0
- item 24 = 39.5
- item 25 = 41.0

In [15]:
quantity_list = retail_sales_data['Quantity'].unique()
print(quantity_list)

[10.  9.  2.  7.  8. nan  1.  3.  6.  4.  5.]


In [16]:
customer_table = retail_sales_data[["Customer ID","Location"]]
customer_table

Customer ID  Location
0         CUST_09    Online
1         CUST_22    Online
2         CUST_02    Online
3         CUST_06    Online
4         CUST_05    Online
...           ...       ...
12570     CUST_18  In-store
12571     CUST_03    Online
12572     CUST_11    Online
12573     CUST_04    Online
12574     CUST_23    Online

[12575 rows x 2 columns]

In [17]:
Transaction_table = retail_sales_data[["Transaction ID", "Customer ID", "Transaction Date", "Payment Method", "Discount Applied"]]
Transaction_table

Transaction ID Customer ID Transaction Date  Payment Method  \
0        TXN_6867343     CUST_09       2024-04-08  Digital Wallet   
1        TXN_3731986     CUST_22       2023-07-23  Digital Wallet   
2        TXN_9303719     CUST_02       2022-10-05     Credit Card   
3        TXN_9458126     CUST_06       2022-05-07     Credit Card   
4        TXN_4575373     CUST_05       2022-10-02  Digital Wallet   
...              ...         ...              ...             ...   
12570    TXN_9347481     CUST_18       2023-09-03     Credit Card   
12571    TXN_4009414     CUST_03       2022-08-12            Cash   
12572    TXN_5306010     CUST_11       2024-08-24            Cash   
12573    TXN_5167298     CUST_04       2023-12-30            Cash   
12574    TXN_2407494     CUST_23       2022-08-06            Cash   

      Discount Applied  
0                 True  
1                 True  
2                False  
3                  NaN  
4                False  
...                ...  
12570              NaN  
12571            False  
12572              NaN  
12573             True  
12574              NaN  

[12575 rows x 5 columns]

In [18]:
Item_table = retail_sales_data[["Category","Item","Price Per Unit"]]
Item_table

Category          Item  Price Per Unit
0         Patisserie   Item_10_PAT            18.5
1      Milk Products  Item_17_MILK            29.0
2           Butchers   Item_12_BUT            21.5
3          Beverages   Item_16_BEV            27.5
4               Food   Item_6_FOOD            12.5
...              ...           ...             ...
12570     Patisserie   Item_23_PAT            38.0
12571      Beverages    Item_2_BEV             6.5
12572       Butchers    Item_7_BUT            14.0
12573      Furniture    Item_7_FUR            14.0
12574           Food   Item_9_FOOD            17.0

[12575 rows x 3 columns]

In [19]:
Sales_table = retail_sales_data[["Transaction ID", "Item", "Quantity", "Total Spent"]]
Sales_table

Transaction ID          Item  Quantity  Total Spent
0        TXN_6867343   Item_10_PAT      10.0        185.0
1        TXN_3731986  Item_17_MILK       9.0        261.0
2        TXN_9303719   Item_12_BUT       2.0         43.0
3        TXN_9458126   Item_16_BEV       9.0        247.5
4        TXN_4575373   Item_6_FOOD       7.0         87.5
...              ...           ...       ...          ...
12570    TXN_9347481   Item_23_PAT       4.0        152.0
12571    TXN_4009414    Item_2_BEV       9.0         58.5
12572    TXN_5306010    Item_7_BUT      10.0        140.0
12573    TXN_5167298    Item_7_FUR       6.0         84.0
12574    TXN_2407494   Item_9_FOOD       3.0         51.0

[12575 rows x 4 columns]

In [20]:
expr = retail_sales_data[["Category","Item","Price Per Unit","Quantity", "Total Spent"]]

In [21]:
expr

Category          Item  Price Per Unit  Quantity  Total Spent
0         Patisserie   Item_10_PAT            18.5      10.0        185.0
1      Milk Products  Item_17_MILK            29.0       9.0        261.0
2           Butchers   Item_12_BUT            21.5       2.0         43.0
3          Beverages   Item_16_BEV            27.5       9.0        247.5
4               Food   Item_6_FOOD            12.5       7.0         87.5
...              ...           ...             ...       ...          ...
12570     Patisserie   Item_23_PAT            38.0       4.0        152.0
12571      Beverages    Item_2_BEV             6.5       9.0         58.5
12572       Butchers    Item_7_BUT            14.0      10.0        140.0
12573      Furniture    Item_7_FUR            14.0       6.0         84.0
12574           Food   Item_9_FOOD            17.0       3.0         51.0

[12575 rows x 5 columns]

In [22]:
df = retail_sales_data.copy()

In [23]:
# Fill missing Price Per Unit values using the formula
retail_sales_data["Price Per Unit"] = retail_sales_data.apply(
    lambda row: row["Total Spent"] / row["Quantity"] if row["Price Per Unit"] == 0 and row["Quantity"] != 0 else row["Price Per Unit"],
    axis=1
)

In [24]:
missing_prices = retail_sales_data[retail_sales_data["Price Per Unit"] == 0]
print(missing_prices)


Empty DataFrame
Columns: [Transaction ID, Customer ID, Category, Item, Price Per Unit, Quantity, Total Spent, Payment Method, Location, Transaction Date, Discount Applied]
Index: []


In [25]:
retail_sales_data["Quantity"] = retail_sales_data.apply(
    lambda row: row["Total Spent"] / row["Price Per Unit"] if row["Quantity"] == 0 and row["Price Per Unit"] != 0 else row["Quantity"], 
    axis=1
)

In [26]:
missing_quantity = retail_sales_data[retail_sales_data["Quantity"]== 0]
print(missing_quantity)

Empty DataFrame
Columns: [Transaction ID, Customer ID, Category, Item, Price Per Unit, Quantity, Total Spent, Payment Method, Location, Transaction Date, Discount Applied]
Index: []


In [27]:
retail_sales_data["Total Spent"] = retail_sales_data.apply(
    lambda row: row["Quantity"] * row["Price Per Unit"] if row["Total Spent"] == 0 and row["Quantity"] != 0 else row["Total Spent"], 
    axis=1
)

In [28]:
missing_Total = retail_sales_data[retail_sales_data["Total Spent"]==0]
print(missing_Total)

Empty DataFrame
Columns: [Transaction ID, Customer ID, Category, Item, Price Per Unit, Quantity, Total Spent, Payment Method, Location, Transaction Date, Discount Applied]
Index: []


In [29]:
retail_sales_data.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [30]:
missing_prices = retail_sales_data[retail_sales_data["Price Per Unit"] == 0]

In [31]:
retail_sales_data

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12570       4.0        152.0     Credit Card  In-store       2023-09-03   
12571       9.0         58.5            Cash    Online       2022-08-12   
12572      10.0        140.0            Cash    Online       2024-08-24   
12573       6.0         84.0            Cash    Online       2023-12-30   
12574       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied  
0                 True  
1                 True  
2                False  
3                  NaN  
4                False  
...                ...  
12570              NaN  
12571            False  
12572              NaN  
12573             True  
12574              NaN  

[12575 rows x 11 columns]

Fillers for the null values in the price, quantity and total spent

In [32]:
#Price Per Unit
df['Price Per Unit'] = df.apply(
    lambda row: row['Total Spent'] / row['Quantity'] if pd.isnull(row['Price Per Unit']) and row['Quantity'] > 0 else row['Price Per Unit'], 
    axis=1
)

# Quantity
df['Quantity'] = df.apply(
    lambda row: row['Total Spent'] / row['Price Per Unit'] if pd.isnull(row['Quantity']) and row['Price Per Unit'] > 0 else row['Quantity'], 
    axis=1
)

# Total Spent
df['Total Spent'] = df.apply(
    lambda row: row['Price Per Unit'] * row['Quantity'] if pd.isnull(row['Total Spent']) and row['Price Per Unit'] > 0 and row['Quantity'] > 0 else row['Total Spent'], 
    axis=1
)

# Replace any remaining nulls
df.fillna({
    'Price Per Unit': 0,
    'Quantity': 0,
    'Total Spent': 0
}, inplace=True)

# Verify
print(df.isnull().sum())


Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit         0
Quantity               0
Total Spent            0
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64


In [33]:
df.sample(n=15)

Transaction ID Customer ID                            Category  \
10244    TXN_7283902     CUST_23                           Beverages   
2369     TXN_5483324     CUST_21                                Food   
4172     TXN_9369677     CUST_19                          Patisserie   
8320     TXN_8048553     CUST_04       Electric household essentials   
8527     TXN_8794922     CUST_06  Computers and electric accessories   
1448     TXN_7686075     CUST_15       Electric household essentials   
6467     TXN_8627287     CUST_21                       Milk Products   
10638    TXN_3023899     CUST_14                           Furniture   
11260    TXN_7402104     CUST_03  Computers and electric accessories   
3598     TXN_8050462     CUST_03                                Food   
606      TXN_9191575     CUST_01                       Milk Products   
4768     TXN_8628627     CUST_16       Electric household essentials   
393      TXN_9671345     CUST_21                           Furniture   
2695     TXN_3433029     CUST_20       Electric household essentials   
9042     TXN_7068871     CUST_11                                Food   

               Item  Price Per Unit  Quantity  Total Spent  Payment Method  \
10244   Item_21_BEV            35.0       5.0        175.0            Cash   
2369   Item_12_FOOD            21.5       8.0        172.0  Digital Wallet   
4172    Item_19_PAT            32.0       4.0        128.0            Cash   
8320     Item_8_EHE            15.5       5.0         77.5            Cash   
8527     Item_5_CEA            11.0       3.0         33.0  Digital Wallet   
1448    Item_16_EHE            27.5       2.0         55.0     Credit Card   
6467   Item_12_MILK            21.5       2.0         43.0            Cash   
10638    Item_5_FUR            11.0       8.0         88.0  Digital Wallet   
11260    Item_5_CEA            11.0       8.0         88.0     Credit Card   
3598   Item_25_FOOD            41.0       3.0        123.0  Digital Wallet   
606     Item_4_MILK             9.5       7.0         66.5  Digital Wallet   
4768    Item_19_EHE            32.0       9.0        288.0     Credit Card   
393     Item_20_FUR            33.5       2.0         67.0            Cash   
2695    Item_15_EHE            26.0       3.0         78.0  Digital Wallet   
9042    Item_9_FOOD            17.0       9.0        153.0            Cash   

       Location Transaction Date Discount Applied  
10244  In-store       2024-06-28            False  
2369     Online       2024-07-17              NaN  
4172     Online       2024-03-05              NaN  
8320   In-store       2024-08-08              NaN  
8527   In-store       2024-05-28            False  
1448     Online       2023-12-31             True  
6467   In-store       2024-01-11              NaN  
10638  In-store       2024-12-30              NaN  
11260    Online       2024-07-31            False  
3598   In-store       2024-05-30            False  
606    In-store       2024-01-28              NaN  
4768     Online       2023-01-18            False  
393      Online       2022-09-13            False  
2695     Online       2025-01-16              NaN  
9042   In-store       2023-02-18            False

Analyze unique combinations of Category and Item

In [34]:

category_item_map = df.groupby("Category")["Item"].unique().reset_index()

print(category_item_map)

                             Category  \
0                           Beverages   
1                            Butchers   
2  Computers and electric accessories   
3       Electric household essentials   
4                                Food   
5                           Furniture   
6                       Milk Products   
7                          Patisserie   

                                                Item  
0  [Item_16_BEV, nan, Item_7_BEV, Item_25_BEV, It...  
1  [Item_12_BUT, Item_22_BUT, Item_3_BUT, Item_20...  
2  [Item_16_CEA, Item_5_CEA, Item_15_CEA, Item_24...  
3  [Item_13_EHE, Item_4_EHE, Item_21_EHE, Item_10...  
4  [Item_6_FOOD, Item_1_FOOD, Item_2_FOOD, Item_1...  
5  [nan, Item_16_FUR, Item_14_FUR, Item_25_FUR, I...  
6  [Item_17_MILK, nan, Item_16_MILK, Item_19_MILK...  
7  [Item_10_PAT, nan, Item_24_PAT, Item_17_PAT, I...  


Display sorted item list for each category, excluding NaN values

In [35]:
for category, items in df.groupby("Category")["Item"].unique().items():
    items = [item for item in items if isinstance(item, str)]  
    sorted_items = sorted(items)  
    print(f"Category: {category}")
    print(f"Items: {sorted_items}")
    print("-" * 40)


Category: Beverages
Items: ['Item_10_BEV', 'Item_11_BEV', 'Item_12_BEV', 'Item_13_BEV', 'Item_14_BEV', 'Item_15_BEV', 'Item_16_BEV', 'Item_17_BEV', 'Item_18_BEV', 'Item_19_BEV', 'Item_1_BEV', 'Item_20_BEV', 'Item_21_BEV', 'Item_22_BEV', 'Item_23_BEV', 'Item_24_BEV', 'Item_25_BEV', 'Item_2_BEV', 'Item_3_BEV', 'Item_4_BEV', 'Item_5_BEV', 'Item_6_BEV', 'Item_7_BEV', 'Item_8_BEV', 'Item_9_BEV']
----------------------------------------
Category: Butchers
Items: ['Item_10_BUT', 'Item_11_BUT', 'Item_12_BUT', 'Item_13_BUT', 'Item_14_BUT', 'Item_15_BUT', 'Item_16_BUT', 'Item_17_BUT', 'Item_18_BUT', 'Item_19_BUT', 'Item_1_BUT', 'Item_20_BUT', 'Item_21_BUT', 'Item_22_BUT', 'Item_23_BUT', 'Item_24_BUT', 'Item_25_BUT', 'Item_2_BUT', 'Item_3_BUT', 'Item_4_BUT', 'Item_5_BUT', 'Item_6_BUT', 'Item_7_BUT', 'Item_8_BUT', 'Item_9_BUT']
----------------------------------------
Category: Computers and electric accessories
Items: ['Item_10_CEA', 'Item_11_CEA', 'Item_12_CEA', 'Item_13_CEA', 'Item_14_CEA', 'It

In [36]:
df['Item'] = df['Item'].fillna('Unknown')

In [37]:
df["Discount Applied"] = df["Discount Applied"].fillna("null")

In [38]:
customers_cat = df.groupby("Customer ID")["Category"]
print(customers_cat)

In [39]:
# Iterate over each group and print
for id, cat in customers_cat:
    print(f"ID: {id}")
    print(len(cat))


ID: CUST_01
507
ID: CUST_02
488
ID: CUST_03
465
ID: CUST_04
474
ID: CUST_05
544
ID: CUST_06
481
ID: CUST_07
491
ID: CUST_08
533
ID: CUST_09
519
ID: CUST_10
501
ID: CUST_11
503
ID: CUST_12
498
ID: CUST_13
534
ID: CUST_14
484
ID: CUST_15
519
ID: CUST_16
515
ID: CUST_17
487
ID: CUST_18
507
ID: CUST_19
487
ID: CUST_20
507
ID: CUST_21
498
ID: CUST_22
501
ID: CUST_23
513
ID: CUST_24
543
ID: CUST_25
476


In [40]:
for id, cat in df.groupby("Customer ID")["Category"].unique().items():
    print(id)
    print(len(cat))

CUST_01
8
CUST_02
8
CUST_03
8
CUST_04
8
CUST_05
8
CUST_06
8
CUST_07
8
CUST_08
8
CUST_09
8
CUST_10
8
CUST_11
8
CUST_12
8
CUST_13
8
CUST_14
8
CUST_15
8
CUST_16
8
CUST_17
8
CUST_18
8
CUST_19
8
CUST_20
8
CUST_21
8
CUST_22
8
CUST_23
8
CUST_24
8
CUST_25
8


In [41]:
grouped = df.groupby("Customer ID")["Item"]

# Iterate over each group and print
for id, items in grouped:
    print(f"Customer ID: {id}")
    print(len(items))


Customer ID: CUST_01
507
Customer ID: CUST_02
488
Customer ID: CUST_03
465
Customer ID: CUST_04
474
Customer ID: CUST_05
544
Customer ID: CUST_06
481
Customer ID: CUST_07
491
Customer ID: CUST_08
533
Customer ID: CUST_09
519
Customer ID: CUST_10
501
Customer ID: CUST_11
503
Customer ID: CUST_12
498
Customer ID: CUST_13
534
Customer ID: CUST_14
484
Customer ID: CUST_15
519
Customer ID: CUST_16
515
Customer ID: CUST_17
487
Customer ID: CUST_18
507
Customer ID: CUST_19
487
Customer ID: CUST_20
507
Customer ID: CUST_21
498
Customer ID: CUST_22
501
Customer ID: CUST_23
513
Customer ID: CUST_24
543
Customer ID: CUST_25
476


#### Some Insights

Some things I found while cleaning the data and from my analysis

- Categories - 8
- Items - 200/8 = 25, 
- which means there are 25 items under each category(8), which makes 200 unique items.
- Price Per unit - between 5 to 41
- CustomerID -- 25(unique)

As you can see, despite the category each item has a unique price, from 5 to 41

- item 1 = 5.0
- item 2 = 6.5
- item 3 = 8.0
- item 4 = 9.5
- item 5 = 11.0
- item 6 = 12.5
- item 7 = 14.0
- item 8 = 15.5
- item 9 = 17.0
- item 10 = 18.5
- item 11 = 20.0
- item 12 = 21.5
- item 13 = 23.0
- item 14 = 24.5
- item 15 = 26.0 
- item 16 = 27.5
- item 17 = 29.0
- item 18 = 30.5
- item 19 = 32.0
- item 20 = 33.5
- item 21 = 35.0
- item 22 = 36.5
- item 23 = 38.0
- item 24 = 39.5
- item 25 = 41.0

#### Feature Engineering
- Introducing new columsn Year, Month and Year-Month

In [42]:

df['Transaction Date'] = pd.to_datetime(df['Transaction Date'])
df['Year'] = df['Transaction Date'].dt.year
df['Month'] = df['Transaction Date'].dt.month
df['Year-Month'] = df['Transaction Date'].dt.to_period('M')

In [43]:
df.sample(n=11)

Transaction ID Customer ID                            Category  \
1360     TXN_9999124     CUST_02                            Butchers   
1406     TXN_5315119     CUST_10                          Patisserie   
4365     TXN_8083639     CUST_14                          Patisserie   
8205     TXN_5504521     CUST_05                           Beverages   
9418     TXN_4351750     CUST_12                                Food   
12353    TXN_8074184     CUST_04                       Milk Products   
12484    TXN_4236200     CUST_15                          Patisserie   
10367    TXN_6968520     CUST_21       Electric household essentials   
11584    TXN_1033656     CUST_18       Electric household essentials   
1603     TXN_8190786     CUST_03  Computers and electric accessories   
5832     TXN_4355077     CUST_24                                Food   

               Item  Price Per Unit  Quantity  Total Spent  Payment Method  \
1360        Unknown            36.5       0.0          0.0  Digital Wallet   
1406        Unknown             9.5       3.0         28.5  Digital Wallet   
4365    Item_17_PAT            29.0       3.0         87.0            Cash   
8205        Unknown            35.0       0.0          0.0  Digital Wallet   
9418   Item_19_FOOD            32.0       4.0        128.0            Cash   
12353  Item_24_MILK            39.5       2.0         79.0            Cash   
12484    Item_3_PAT             8.0       3.0         24.0     Credit Card   
10367   Item_20_EHE            33.5       7.0        234.5     Credit Card   
11584   Item_15_EHE            26.0       5.0        130.0     Credit Card   
1603    Item_20_CEA            33.5       4.0        134.0            Cash   
5832        Unknown            29.0       5.0        145.0            Cash   

       Location Transaction Date Discount Applied  Year  Month Year-Month  
1360   In-store       2024-02-21            False  2024      2    2024-02  
1406     Online       2024-09-01             null  2024      9    2024-09  
4365     Online       2023-08-03             null  2023      8    2023-08  
8205   In-store       2024-04-19            False  2024      4    2024-04  
9418   In-store       2022-11-04             True  2022     11    2022-11  
12353    Online       2024-05-14            False  2024      5    2024-05  
12484  In-store       2023-09-01             null  2023      9    2023-09  
10367    Online       2024-01-27             True  2024      1    2024-01  
11584    Online       2024-04-29             null  2024      4    2024-04  
1603     Online       2022-02-11             null  2022      2    2022-02  
5832   In-store       2023-02-24             null  2023      2    2023-02

Saving the final Data frame into CSV format

In [44]:
df.to_csv("visual_retail_sales_data.csv", index=False)